In [1]:
import os
os.chdir("D:/UIT/HK/HK6/IS403/Đồ án/Source code")
print("Thư mục làm việc hiện tại:", os.getcwd())

Thư mục làm việc hiện tại: D:\UIT\HK\HK6\IS403\Đồ án\Source code


# Import modules

In [2]:
import pandas as pd

In [3]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# Load data

In [4]:
bid_kqkd = pd.read_csv("./dataset/clean_dataset/bank_index/bid/bid_kqkd.csv")
bid_kqkd

Năm  Kỳ  Doanh thu (đồng)  Tăng trưởng doanh thu (%)  \
0   2024   4    36469243000000                  -0.027386   
1   2024   3    34326293000000                  -0.099149   
2   2024   2    33896893000000                  -0.140234   
3   2024   1    33595982000000                  -0.109691   
4   2023   4    37496112000000                   0.103046   
5   2023   3    38104286000000                   0.244256   
6   2023   2    39425710000000                   0.339555   
7   2023   1    37735207000000                   0.400121   
8   2022   4    33993242000000                   0.371351   
9   2022   3    30624151000000                   0.178320   
10  2022   2    29431951000000                   0.124599   
11  2022   1    26951389000000                   0.120256   
12  2021   4    24788144000000                  -0.045132   

    Lợi nhuận sau thuế của Cổ đông công ty mẹ (đồng)  \
0                                      7810117000000   
1                                      5151318000000   
2                                      6365954000000   
3                                      5812648000000   
4                                      6074285000000   
5                                      4582880000000   
6                                      5426134000000   
7                                      5468480000000   
8                                      4078937000000   
9                                      5277956000000   
10                                     5157859000000   
11                                     3571160000000   
12                                     2217704000000   

    Tăng trưởng lợi nhuận (%)  LN trước thuế  Lợi nhuận thuần  Cố tức đã nhận  
0                    0.285767  9938144000000    7946423000000    144692000000  
1                    0.124035  6498279000000    5213866000000    105983000000  
2                    0.173203  8158532000000    6528169000000     99423000000  
3                    0.062937  7389914000000    5915579000000     95644000000  
4                    0.489183  7886764000000    6190625000000     75128000000  
5                   -0.131694  5892788000000    4727935000000     96319000000  
6                    0.052013  6950633000000    5549837000000     73920000000  
7                    0.531290  6919579000000    5559041000000    109908000000  
8                    0.839261  5245911000000    4156992000000     44232000000  
9                    1.577107  6673383000000    5354232000000     99667000000  
10                   0.409747  6489854000000    5199635000000     63616000000  
11                   0.348556  4513501000000    3637665000000     83437000000  
12                   0.355021  2868211000000    2295616000000     34547000000

In [5]:
bid_cstc = pd.read_csv("./dataset/clean_dataset/bank_index/bid/bid_cstc.csv")
bid_cstc

CP   Năm  Kỳ  Biên lợi nhuận ròng (%)   ROE (%)  ROIC (%)   ROA (%)  \
0   BID  2024   4                 0.499409  0.187768       0.0  0.009934   
1   BID  2024   3                 0.368222  0.178287       0.0  0.009625   
2   BID  2024   2                 0.429037  0.180613       0.0  0.009839   
3   BID  2024   1                 0.429251  0.180520       0.0  0.009855   
4   BID  2023   4                 0.408518  0.189483       0.0  0.009727   
5   BID  2023   3                 0.332500  0.175848       0.0  0.009270   
6   BID  2023   2                 0.400536  0.189769       0.0  0.009693   
7   BID  2023   1                 0.392394  0.196536       0.0  0.009734   
8   BID  2022   4                 0.282706  0.190621       0.0  0.009354   
9   BID  2022   3                 0.374365  0.174907       0.0  0.008497   
10  BID  2022   2                 0.353105  0.145951       0.0  0.007144   
11  BID  2022   1                 0.278433  0.132601       0.0  0.006627   
12  BID  2021   4                 0.204334  0.127008       0.0  0.006430   

    Tỷ suất cổ tức (%)  Vốn hóa (Tỷ đồng)  Số CP lưu hành (Triệu CP)  \
0                  0.0    256279709970500                 7021361917   
1                  0.0    269003095452000                 6897515268   
2                  0.0    281601533460000                 5700435900   
3                  0.0    265640312940000                 5700435900   
4                  0.0    304403277060000                 5700435900   
5                  0.0    262220051400000                 5700435900   
6                  0.0    213216778844400                 5058523816   
7                  0.0    236233062207200                 5058523816   
8                  0.0    224851383621200                 5058523816   
9                  0.0    211446295508800                 5058523816   
10                 0.0    164149097829200                 5058523816   
11                 0.0    182359783566800                 5058523816   
12                 0.0    188430012146000                 5058523816   

          P/E       P/B       P/S  P/Cash Flow    EPS (VND)    BVPS (VND)  
0   10.037253  1.773779  3.161571     2.552809  1112.336480  20577.535482  
1   11.493793  2.017273  3.574493     4.012717   746.836767  19333.031073  
2   12.331600  2.171694  3.710919     3.893225  1116.748633  22747.219208  
3   12.131940  2.142925  3.643057    15.141780  1019.684828  21745.973497  
4   14.155054  2.581861  4.169164    -4.286465  1065.582546  20682.757787  
5   13.335113  2.295202  3.740526    -4.587722   803.952554  20041.812066  
6   10.472875  1.949019  3.055989   -15.461085  1072.671435  21626.268449  
7   11.758341  2.249375  3.343974     7.025580  1081.042652  20761.322239  
8   12.382706  2.260543  3.231438     1.696762   806.349273  19663.415577  
9   13.032387  2.196679  3.148185     1.881048  1043.378699  19028.722114  
10  12.631966  1.805028  2.530948     2.033995  1019.637188  17977.563279  
11  15.863453  2.105010  2.848616     5.630432   705.968802  17125.805897  
12  17.877376  2.266552  3.015179     2.735225   438.409323  16434.655845

In [6]:
bid_cdtc = pd.read_csv("./dataset/clean_dataset/bank_index/bid/bid_cdtc.csv")
bid_cdtc

Năm  Kỳ  TỔNG CỘNG TÀI SẢN (đồng)  TỔNG CỘNG NGUỒN VỐN (đồng)  \
0   2024   4          2760791854000000            2760791854000000   
1   2024   3          2575981275000000            2575981275000000   
2   2024   2          2521071363000000            2521071363000000   
3   2024   1          2328730752000000            2328730752000000   
4   2023   4          2300814314000000            2300814314000000   
5   2023   3          2132854577000000            2132854577000000   
6   2023   2          2124767632000000            2124767632000000   
7   2023   1          2106864549000000            2106864549000000   
8   2022   4          2120527692000000            2120527692000000   
9   2022   3          2048952718000000            2048952718000000   
10  2022   2          1979292691000000            1979292691000000   
11  2022   1          1847704434000000            1847704434000000   
12  2021   4          1761938950000000            1761938950000000   

    NỢ PHẢI TRẢ (đồng)  VỐN CHỦ SỞ HỮU (đồng)  LỢI ÍCH CỦA CỔ ĐÔNG THIỂU SỐ  
0     2615880741000000        144911113000000                 5234039000000  
1     2437541619000000        138439656000000                 5089779000000  
2     2386354090000000        134717273000000                 5048208000000  
3     2199712917000000        129017835000000                 5056307000000  
4     2177897270000000        122917044000000                 4966455000000  
5     2013766739000000        119087838000000                 4840773000000  
6     2010615988000000        114151644000000                 4754650000000  
7     1997016087000000        109848462000000                 4826819000000  
8     2016322127000000        104205565000000                 4719327000000  
9     1948027658000000        100925060000000                 4667816000000  
10    1885375152000000         93917539000000                 2977607000000  
11    1757866709000000         89837725000000                 3206428000000  
12    1675572305000000         86366645000000                 3184070000000

# Perform descriptive statistics

In [7]:
bid_kqkd.describe()

Năm         Kỳ  Doanh thu (đồng)  Tăng trưởng doanh thu (%)  \
count    13.000000  13.000000      1.300000e+01                  13.000000   
mean   2022.846154   2.615385      3.360297e+13                   0.112301   
std       0.987096   1.192928      4.505469e+12                   0.188354   
min    2021.000000   1.000000      2.478814e+13                  -0.140234   
25%    2022.000000   2.000000      3.062415e+13                  -0.045132   
50%    2023.000000   3.000000      3.399324e+13                   0.120256   
75%    2024.000000   4.000000      3.749611e+13                   0.244256   
max    2024.000000   4.000000      3.942571e+13                   0.400121   

       Lợi nhuận sau thuế của Cổ đông công ty mẹ (đồng)  \
count                                      1.300000e+01   
mean                                       5.153495e+12   
std                                        1.372407e+12   
min                                        2.217704e+12   
25%                                        4.582880e+12   
50%                                        5.277956e+12   
75%                                        5.812648e+12   
max                                        7.810117e+12   

       Tăng trưởng lợi nhuận (%)  LN trước thuế  Lợi nhuận thuần  \
count                  13.000000   1.300000e+01     1.300000e+01   
mean                    0.393571   6.571192e+12     5.251970e+12   
std                     0.434549   1.748989e+12     1.391267e+12   
min                    -0.131694   2.868211e+12     2.295616e+12   
25%                     0.124035   5.892788e+12     4.727935e+12   
50%                     0.348556   6.673383e+12     5.354232e+12   
75%                     0.489183   7.389914e+12     5.915579e+12   
max                     1.577107   9.938144e+12     7.946423e+12   

       Cố tức đã nhận  
count    1.300000e+01  
mean     8.665508e+10  
std      2.906925e+10  
min      3.454700e+10  
25%      7.392000e+10  
50%      9.564400e+10  
75%      9.966700e+10  
max      1.446920e+11

In [8]:
bid_cdtc.describe()

Năm         Kỳ  TỔNG CỘNG TÀI SẢN (đồng)  \
count    13.000000  13.000000              1.300000e+01   
mean   2022.846154   2.615385              2.200792e+15   
std       0.987096   1.192928              2.884381e+14   
min    2021.000000   1.000000              1.761939e+15   
25%    2022.000000   2.000000              2.048953e+15   
50%    2023.000000   3.000000              2.124768e+15   
75%    2024.000000   4.000000              2.328731e+15   
max    2024.000000   4.000000              2.760792e+15   

       TỔNG CỘNG NGUỒN VỐN (đồng)  NỢ PHẢI TRẢ (đồng)  VỐN CHỦ SỞ HỮU (đồng)  \
count                1.300000e+01        1.300000e+01           1.300000e+01   
mean                 2.200792e+15        2.086304e+15           1.144880e+14   
std                  2.884381e+14        2.700347e+14           1.902784e+13   
min                  1.761939e+15        1.675572e+15           8.636664e+13   
25%                  2.048953e+15        1.948028e+15           1.009251e+14   
50%                  2.124768e+15        2.013767e+15           1.141516e+14   
75%                  2.328731e+15        2.199713e+15           1.290178e+14   
max                  2.760792e+15        2.615881e+15           1.449111e+14   

       LỢI ÍCH CỦA CỔ ĐÔNG THIỂU SỐ  
count                  1.300000e+01  
mean                   4.505560e+12  
std                    8.062553e+11  
min                    2.977607e+12  
25%                    4.667816e+12  
50%                    4.826819e+12  
75%                    5.048208e+12  
max                    5.234039e+12

In [9]:
bid_cstc.describe()

Năm         Kỳ  Biên lợi nhuận ròng (%)    ROE (%)  ROIC (%)  \
count    13.000000  13.000000                13.000000  13.000000      13.0   
mean   2022.846154   2.615385                 0.365601   0.173070       0.0   
std       0.987096   1.192928                 0.077229   0.022856       0.0   
min    2021.000000   1.000000                 0.204334   0.127008       0.0   
25%    2022.000000   2.000000                 0.332500   0.174907       0.0   
50%    2023.000000   3.000000                 0.374365   0.180520       0.0   
75%    2024.000000   4.000000                 0.408518   0.189483       0.0   
max    2024.000000   4.000000                 0.499409   0.196536       0.0   

         ROA (%)  Tỷ suất cổ tức (%)  Vốn hóa (Tỷ đồng)  \
count  13.000000                13.0       1.300000e+01   
mean    0.008902                 0.0       2.353719e+14   
std     0.001299                 0.0       4.219886e+13   
min     0.006430                 0.0       1.641491e+14   
25%     0.008497                 0.0       2.114463e+14   
50%     0.009625                 0.0       2.362331e+14   
75%     0.009734                 0.0       2.656403e+14   
max     0.009934                 0.0       3.044033e+14   

       Số CP lưu hành (Triệu CP)        P/E        P/B        P/S  \
count               1.300000e+01  13.000000  13.000000  13.000000   
mean                5.548484e+09  12.884912   2.139611   3.321082   
std                 6.929429e+08   2.120434   0.217015   0.436875   
min                 5.058524e+09  10.037253   1.773779   2.530948   
25%                 5.058524e+09  11.758341   2.017273   3.055989   
50%                 5.058524e+09  12.382706   2.171694   3.231438   
75%                 5.700436e+09  13.335113   2.260543   3.643057   
max                 7.021362e+09  17.877376   2.581861   4.169164   

       P/Cash Flow    EPS (VND)    BVPS (VND)  
count    13.000000    13.000000     13.000000  
mean      1.712946   925.584552  19826.621732  
std       7.109076   207.149376   1843.798370  
min     -15.461085   438.409323  16434.655845  
25%       1.696762   803.952554  19028.722114  
50%       2.552809  1019.684828  20041.812066  
75%       4.012717  1072.671435  20761.322239  
max      15.141780  1116.748633  22747.219208

# Visualize

## Kết quả kinh doanh

In [10]:
bid_kqkd["Tháng"] = bid_kqkd["Kỳ"] * 3
bid_kqkd

Năm  Kỳ  Doanh thu (đồng)  Tăng trưởng doanh thu (%)  \
0   2024   4    36469243000000                  -0.027386   
1   2024   3    34326293000000                  -0.099149   
2   2024   2    33896893000000                  -0.140234   
3   2024   1    33595982000000                  -0.109691   
4   2023   4    37496112000000                   0.103046   
5   2023   3    38104286000000                   0.244256   
6   2023   2    39425710000000                   0.339555   
7   2023   1    37735207000000                   0.400121   
8   2022   4    33993242000000                   0.371351   
9   2022   3    30624151000000                   0.178320   
10  2022   2    29431951000000                   0.124599   
11  2022   1    26951389000000                   0.120256   
12  2021   4    24788144000000                  -0.045132   

    Lợi nhuận sau thuế của Cổ đông công ty mẹ (đồng)  \
0                                      7810117000000   
1                                      5151318000000   
2                                      6365954000000   
3                                      5812648000000   
4                                      6074285000000   
5                                      4582880000000   
6                                      5426134000000   
7                                      5468480000000   
8                                      4078937000000   
9                                      5277956000000   
10                                     5157859000000   
11                                     3571160000000   
12                                     2217704000000   

    Tăng trưởng lợi nhuận (%)  LN trước thuế  Lợi nhuận thuần  Cố tức đã nhận  \
0                    0.285767  9938144000000    7946423000000    144692000000   
1                    0.124035  6498279000000    5213866000000    105983000000   
2                    0.173203  8158532000000    6528169000000     99423000000   
3                    0.062937  7389914000000    5915579000000     95644000000   
4                    0.489183  7886764000000    6190625000000     75128000000   
5                   -0.131694  5892788000000    4727935000000     96319000000   
6                    0.052013  6950633000000    5549837000000     73920000000   
7                    0.531290  6919579000000    5559041000000    109908000000   
8                    0.839261  5245911000000    4156992000000     44232000000   
9                    1.577107  6673383000000    5354232000000     99667000000   
10                   0.409747  6489854000000    5199635000000     63616000000   
11                   0.348556  4513501000000    3637665000000     83437000000   
12                   0.355021  2868211000000    2295616000000     34547000000   

    Tháng  
0      12  
1       9  
2       6  
3       3  
4      12  
5       9  
6       6  
7       3  
8      12  
9       9  
10      6  
11      3  
12     12

In [11]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=bid_kqkd["Năm"].astype(str) + "-" + bid_kqkd["Tháng"].astype(str),
    y=bid_kqkd["Doanh thu (đồng)"],
    mode="lines+markers",
    name="Doanh thu (đồng)"
))

fig.update_layout(
    title="Doanh thu (đồng) theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Doanh thu (đồng)"
)

fig.show()


Theo biểu đồ, ta thấy:
- Từ đầu năm 2022 - 2025, doanh thu có xu hướng tăng và đạt đỉnh ở quý 2 năm 2023. Giai đoạn tăng mạnh nhất từ quý 2 2022 - quý 1 năm 2023.
- Sau quý 2 năm 2023, doanh thu có xu hướng giảm nhẹ. Nhưng từ đầu năm 2024, doanh thu bắt đầu phục hồi.

In [12]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=bid_kqkd["Năm"].astype(str) + "-" + bid_kqkd["Tháng"].astype(str),
    y=bid_kqkd["Tăng trưởng doanh thu (%)"],
    marker_color=["green" if val > 0 else "red" for val in bid_kqkd["Tăng trưởng doanh thu (%)"]]
))

fig.update_layout(
    title="Tăng trưởng doanh thu (%) theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Tăng trưởng doanh thu (%)"
)

fig.show()


Theo biểu đồ tăng trưởng doanh thu, ta thấy:
- Từ đầu năm 2022 - 2023, tăng trưởng doanh thu luôn dương. Trong đó, quý 2 năm 2023 tăng trưởng doanh thu đạt đỉnh.
- Trong năm 2022 - quý 2 2023, tăng trưởng doanh thu cao nhưng ổn định.
- Từ quý 2 năm 2023, tăng trưởng doanh thu có xu hướng giảm. Và bắt đầu tăng trưởng âm vào năm 2024, trong đó đạt đáy vào quý 2 2024. Sau đó tăng trưởng doanh thu có dấu hiệu phục hồi.

In [13]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=bid_kqkd["Năm"].astype(str) + "-" + bid_kqkd["Tháng"].astype(str),
    y=bid_kqkd["Tăng trưởng lợi nhuận (%)"],
    marker_color=["green" if val > 0 else "red" for val in bid_kqkd["Tăng trưởng lợi nhuận (%)"]]
))

fig.update_layout(
    title="Tăng trưởng lợi nhuận (%) theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Tăng trưởng lợi nhuận (%)"
)

fig.show()


Theo biểu đồ tăng trưởng lợi nhuận, ta thấy:
- Từ năm 2022 đến quý 3 năm 2023, tăng trưởng lợi nhuận dương, trong đó quý 4 năm 2022 tăng trưởng đạt đỉnh.
- Tăng trưởng lợi nhuận từ năm 2023 có xu hướng giảm, và tăng trưởng âm vào quý 3 2023.
- Tăng trưởng lợi nhuận từ cuối năm 2023 - 2024 không ổn định và không có xu hướng rõ ràng.

In [14]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=bid_kqkd["Năm"].astype(str) + "-" + bid_kqkd["Tháng"].astype(str),
    y=bid_kqkd["LN trước thuế"],
    mode="lines+markers",
    name="Lợi nhuận trước thuế (đồng)"
))
fig.add_trace(go.Scatter(
    x=bid_kqkd["Năm"].astype(str) + "-" + bid_kqkd["Tháng"].astype(str),
    y=bid_kqkd["Lợi nhuận thuần"],
    mode="lines+markers",
    name="Lợi nhuận thuần (đồng)"
))

fig.update_layout(
    title="Lợi nhuận trước thuế và Lợi nhuận thuần theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Giá trị (đồng)"
)

fig.show()


Theo biểu đồ, ta thấy:
- Năm 2022 -2024, cả lợi nhuận trước thuế và sau thuế đều có xu hướng tăng, và có tính chu kì (tăng vào đầu năm và sụt giảm vào cuối năm).
- Năm 2023, lợi nhuận trước thuế và sau thuế ổn định quanh mức 4-5 nghìn tỉ đồng.
- Năm 2024, lợi nhuận trước thuế và sau thuế có biến động mạnh và đạt đỉnh vào cuối năm 2024. 

In [15]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=bid_kqkd["Năm"].astype(str) + "-" + bid_kqkd["Tháng"].astype(str),
    y=bid_kqkd["Cố tức đã nhận"],
    name="Cổ tức đã nhận"
))

fig.update_layout(
    title="Cổ tức đã nhận theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Cổ tức đã nhận",
    xaxis_tickangle=-45
)

fig.show()


Theo biểu đồ, ta thấy cổ tức không ổn định, nhưng trong các giai đoạn đều có phân bổ cổ tức.

## Cân đối kế toán

In [16]:
bid_cdtc["Tháng"] = bid_cdtc["Kỳ"] * 3
bid_cdtc

Năm  Kỳ  TỔNG CỘNG TÀI SẢN (đồng)  TỔNG CỘNG NGUỒN VỐN (đồng)  \
0   2024   4          2760791854000000            2760791854000000   
1   2024   3          2575981275000000            2575981275000000   
2   2024   2          2521071363000000            2521071363000000   
3   2024   1          2328730752000000            2328730752000000   
4   2023   4          2300814314000000            2300814314000000   
5   2023   3          2132854577000000            2132854577000000   
6   2023   2          2124767632000000            2124767632000000   
7   2023   1          2106864549000000            2106864549000000   
8   2022   4          2120527692000000            2120527692000000   
9   2022   3          2048952718000000            2048952718000000   
10  2022   2          1979292691000000            1979292691000000   
11  2022   1          1847704434000000            1847704434000000   
12  2021   4          1761938950000000            1761938950000000   

    NỢ PHẢI TRẢ (đồng)  VỐN CHỦ SỞ HỮU (đồng)  LỢI ÍCH CỦA CỔ ĐÔNG THIỂU SỐ  \
0     2615880741000000        144911113000000                 5234039000000   
1     2437541619000000        138439656000000                 5089779000000   
2     2386354090000000        134717273000000                 5048208000000   
3     2199712917000000        129017835000000                 5056307000000   
4     2177897270000000        122917044000000                 4966455000000   
5     2013766739000000        119087838000000                 4840773000000   
6     2010615988000000        114151644000000                 4754650000000   
7     1997016087000000        109848462000000                 4826819000000   
8     2016322127000000        104205565000000                 4719327000000   
9     1948027658000000        100925060000000                 4667816000000   
10    1885375152000000         93917539000000                 2977607000000   
11    1757866709000000         89837725000000                 3206428000000   
12    1675572305000000         86366645000000                 3184070000000   

    Tháng  
0      12  
1       9  
2       6  
3       3  
4      12  
5       9  
6       6  
7       3  
8      12  
9       9  
10      6  
11      3  
12     12

In [17]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=bid_cdtc["Năm"].astype(str) + "-" + bid_cdtc["Tháng"].astype(str),
    y=bid_cdtc["TỔNG CỘNG TÀI SẢN (đồng)"],
    mode="lines+markers",
    name="Tổng cộng tài sản (đồng)"
))

fig.add_trace(go.Scatter(x=bid_cdtc["Năm"].astype(str) + "-" + bid_cdtc["Tháng"].astype(str),
                         y=bid_cdtc["NỢ PHẢI TRẢ (đồng)"],
                         mode="lines+markers",
                         name="Nợ phải trả (đồng)"))

fig.update_layout(title="Tổng cộng tài sản và nợ phải trả theo thời gian",
                  xaxis_title="Năm - Kỳ",
                  yaxis_title="Giá trị (đồng)")

fig.show()


Theo biểu đồ, ta thấy:
- Tổng cộng tài sản và nợ phải trả có xu hướng tăng dần.
- Vốn chủ sở hữu tăng dần theo thời gian.
- Lợi ích cổ đông thiểu số tăng dần theo thời gian. Đặc biệt vào quý 3 2022, có sự tăng trưởng bất thường.

In [18]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=bid_cdtc["Năm"].astype(str) + "-" + bid_cdtc["Tháng"].astype(str),
    y=bid_cdtc["VỐN CHỦ SỞ HỮU (đồng)"],
    mode="lines+markers",
    name="Vốn chủ sở hữu (đồng)"
))

fig.update_layout(
    title="Vốn chủ sở hữu theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Vốn chủ sở hữu (đồng)"
)

fig.show()


In [19]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=bid_cdtc["Năm"].astype(str) + "-" + bid_cdtc["Tháng"].astype(str),
    y=bid_cdtc["LỢI ÍCH CỦA CỔ ĐÔNG THIỂU SỐ"],
    mode="lines+markers",
    name="Lợi ích của cổ đông thiểu số"
))

fig.update_layout(
    title="Lợi ích của cổ đông thiểu số theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Lợi ích của cổ đông thiểu số"
)

fig.show()


## Chỉ số tài chính

In [20]:
bid_cstc["Tháng"] = bid_cstc["Kỳ"] * 3
bid_cstc

CP   Năm  Kỳ  Biên lợi nhuận ròng (%)   ROE (%)  ROIC (%)   ROA (%)  \
0   BID  2024   4                 0.499409  0.187768       0.0  0.009934   
1   BID  2024   3                 0.368222  0.178287       0.0  0.009625   
2   BID  2024   2                 0.429037  0.180613       0.0  0.009839   
3   BID  2024   1                 0.429251  0.180520       0.0  0.009855   
4   BID  2023   4                 0.408518  0.189483       0.0  0.009727   
5   BID  2023   3                 0.332500  0.175848       0.0  0.009270   
6   BID  2023   2                 0.400536  0.189769       0.0  0.009693   
7   BID  2023   1                 0.392394  0.196536       0.0  0.009734   
8   BID  2022   4                 0.282706  0.190621       0.0  0.009354   
9   BID  2022   3                 0.374365  0.174907       0.0  0.008497   
10  BID  2022   2                 0.353105  0.145951       0.0  0.007144   
11  BID  2022   1                 0.278433  0.132601       0.0  0.006627   
12  BID  2021   4                 0.204334  0.127008       0.0  0.006430   

    Tỷ suất cổ tức (%)  Vốn hóa (Tỷ đồng)  Số CP lưu hành (Triệu CP)  \
0                  0.0    256279709970500                 7021361917   
1                  0.0    269003095452000                 6897515268   
2                  0.0    281601533460000                 5700435900   
3                  0.0    265640312940000                 5700435900   
4                  0.0    304403277060000                 5700435900   
5                  0.0    262220051400000                 5700435900   
6                  0.0    213216778844400                 5058523816   
7                  0.0    236233062207200                 5058523816   
8                  0.0    224851383621200                 5058523816   
9                  0.0    211446295508800                 5058523816   
10                 0.0    164149097829200                 5058523816   
11                 0.0    182359783566800                 5058523816   
12                 0.0    188430012146000                 5058523816   

          P/E       P/B       P/S  P/Cash Flow    EPS (VND)    BVPS (VND)  \
0   10.037253  1.773779  3.161571     2.552809  1112.336480  20577.535482   
1   11.493793  2.017273  3.574493     4.012717   746.836767  19333.031073   
2   12.331600  2.171694  3.710919     3.893225  1116.748633  22747.219208   
3   12.131940  2.142925  3.643057    15.141780  1019.684828  21745.973497   
4   14.155054  2.581861  4.169164    -4.286465  1065.582546  20682.757787   
5   13.335113  2.295202  3.740526    -4.587722   803.952554  20041.812066   
6   10.472875  1.949019  3.055989   -15.461085  1072.671435  21626.268449   
7   11.758341  2.249375  3.343974     7.025580  1081.042652  20761.322239   
8   12.382706  2.260543  3.231438     1.696762   806.349273  19663.415577   
9   13.032387  2.196679  3.148185     1.881048  1043.378699  19028.722114   
10  12.631966  1.805028  2.530948     2.033995  1019.637188  17977.563279   
11  15.863453  2.105010  2.848616     5.630432   705.968802  17125.805897   
12  17.877376  2.266552  3.015179     2.735225   438.409323  16434.655845   

    Tháng  
0      12  
1       9  
2       6  
3       3  
4      12  
5       9  
6       6  
7       3  
8      12  
9       9  
10      6  
11      3  
12     12

In [22]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=bid_cstc["Năm"].astype(str) + "-" + bid_cstc["Tháng"].astype(str),
    y=bid_cstc["ROE (%)"],
    mode="lines+markers",
    name="ROE (%)",
    yaxis="y1"
))
fig.add_trace(go.Scatter(
    x=bid_cstc["Năm"].astype(str) + "-" + bid_cstc["Tháng"].astype(str),
    y=bid_cstc["ROA (%)"],
    mode="lines+markers",
    name="ROA (%)",
    yaxis="y2"
))

fig.update_layout(
    title="ROE và ROA theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis=dict(title="ROE (%)", tickfont=dict(color="#1f77b4")),
    yaxis2=dict(title="ROA (%)", overlaying="y", side="right", tickfont=dict(color="#ff7f0e"))
)

fig.show()


Theo biểu đồ, ta thấy đối với ROE:
- Từ năm 2022 - 2023, ROE tăng mạnh.
- Trong năm 2023: ROE có xu hướng giảm dần và đi ngang ở năm 2024, đặc biệt là từ giữa năm 2024 trở đi ROE duy trì ở mức khá cao, dao động quanh 0.19 - 0.20.

Đối với ROA:
- Năm 2022 - giữa năm 2023 ROA tăng mạnh, sau đó có sự điều chỉnh nhẹ ở quý 3 năm 2023 rồi quay lại mức cũ.
- Năm 2024: ROA duy trì ở mức tương đối ổn định quanh 0.01.

Có sự chênh lệch không lớn giữa ROA và ROE cho thấy ngân hàng đang sử dụng đòn bẩy tài chính ở mức an toàn.


In [23]:
fig = go.Figure(data=[go.Bar(
    x=bid_cstc["Năm"].astype(str) + "-" + bid_cstc["Tháng"].astype(str),
    y=bid_cstc["Biên lợi nhuận ròng (%)"],
    name="Biên lợi nhuận ròng (%)"
)])

fig.update_layout(
    title="Biên lợi nhuận ròng (%) theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Biên lợi nhuận ròng (%)",
    xaxis_tickangle=-45
)

fig.show()


Theo biểu đồ, ta thấy:
- Năm 2022: Biên lợi nhuận ròng có xu hướng tăng, tuy nhiên vào quý 4 năm 2022 biên lợi nhuận có sự suy giảm.
- Năm 2023 - 2024: Biên lợi nhuận ổn định quanh mức 0.4, tuy có sự suy giảm vào quý 3 của các năm. Nhưng biên lợi nhuận đã tăng và đạt mức cao nhất là vào cuối năm 2024 ở mức 0.5.

In [24]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=bid_cstc["Năm"].astype(str) + "-" + bid_cstc["Tháng"].astype(str),
    y=bid_cstc["Vốn hóa (Tỷ đồng)"],
    mode="lines+markers",
    name="Vốn hóa (tỷ đồng)"
))

fig.update_layout(
    title="Vốn hóa (tỷ đồng) theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Vốn hóa (tỷ đồng)"
)

fig.show()

Theo biểu đồ, ta thấy:
- Quý 1 năm 2022 - Quý 2 năm 2022: Vốn hóa giảm mạnh từ khoảng 126 nghìn tỷ đồng xuống mức thấp nhất là khoảng 89 nghìn tỷ đồng. Đây là giai đoạn giảm sâu nhất trong toàn bộ chu kỳ.
- Từ sau quý 2 năm 2022: Vốn hoá có xu hướng tăng. Trong đó giai đoạn từ quý 2 2022 - quý 3 2023, vốn hoá tăng đều từ khoảng 89 nghìn tỷ lên 122 nghìn tỷ. Từ sau quý 3 2023 - quý 4 2024, vốn hoá tăng mạnh và có biến động mạnh.

In [ ]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=bid_cstc["Năm"].astype(str) + "-" + bid_cstc["Tháng"].astype(str),
    y=bid_cstc["Số CP lưu hành (Triệu CP)"],
    name="Số cổ phần lưu hành (triệu cổ phần)"
))

fig.update_layout(
    title="Số cổ phần lưu hành (triệu cổ phần) theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Số cổ phần lưu hành (triệu cổ phần)",
    xaxis_tickangle=-45
)

fig.show()

Từ biểu đồ, ta thấy số cổ phiếu lưu hành có 2 lần điều chỉnh vào quý 3 năm 2023 và quý 3 năm 2024.

In [27]:
fig = make_subplots(specs=[[{"secondary_y": True}]])

fig.add_trace(
    go.Bar(
        x=bid_cstc["Năm"].astype(str) + "-" + bid_cstc["Tháng"].astype(str),
        y=bid_cstc["BVPS (VND)"],
        name="BVPS"
    ),
    secondary_y=False
)
fig.add_trace(
    go.Scatter(
        x=bid_cstc["Năm"].astype(str) + "-" + bid_cstc["Tháng"].astype(str),
        y=bid_cstc["EPS (VND)"],
        mode="lines+markers",
        name="EPS"
    ),
    secondary_y=True
)

fig.update_layout(
    title="BVPS và EPS theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="BVPS",
    yaxis2=dict(title="EPS", overlaying="y", side="right")
)

fig.show()

Theo biểu đồ, ta thấy:

Về BVPS:
- BVPS có xu hướng tăng trưởng khá ổn định và liên tục trong suốt giai đoạn khảo sát. 
- Quý 1/2022 - Quý 4/2023: BVPS có xu hướng tăng trưởng khá ổn định trong giai đoạn này. Từ khoảng 28 nghìn đồng vào đầu năm 2022, BVPS tăng dần qua các quý và đạt mức cao nhất khoảng 37 nghìn đồng vào cuối năm 2023.
- Quý 1/2024 - Quý 4/2025: BVPS giảm xuống mức 19 nghìn đồng, sau đó có dấu hiệu đi ngang và duy trì ở mức tương đương, dao động nhẹ quanh mốc 20 nghìn đồng.


Về EPS:
- Năm 2022: EPS có sự biến động lớn trong giai đoạn này. Ban đầu ở mức khá cao, EPS giảm mạnh xuống mức thấp nhất vào cuối năm 2022.
- Năm 2023: EPS cho thấy sự phục hồi rõ rệt và có xu hướng ổn định hơn, dao động quanh mức 1250-1300.
- Từ đầu năm 2024, EPS bắt đầu có xu hướng giảm dần và giảm mạnh vào cuối năm 2024.

So sánh BVPS và EPS:
- BVPS cho thấy sự tăng trưởng ổn định trong phần lớn giai đoạn, cho thấy giá trị tài sản ròng trên mỗi cổ phiếu của công ty đang tăng lên.
- EPS lại có sự biến động mạnh hơn. Giai đoạn đầu có sự sụt giảm đáng kể, sau đó phục hồi nhưng lại suy giảm mạnh vào cuối chu kỳ. Điều này cho thấy lợi nhuận trên mỗi cổ phiếu của công ty không ổn định và có dấu hiệu suy yếu trong giai đoạn gần đây.
- Sự khác biệt trong xu hướng của BVPS và EPS có thể cho thấy công ty đang tích lũy tài sản nhưng hiệu quả sinh lời trên tài sản đó lại giảm sút.


In [29]:
fig = make_subplots(
    rows=3,
    cols=1,
    shared_xaxes=True,
    vertical_spacing=0.03,
    subplot_titles=("P/E theo thời gian", "P/B theo thời gian", "P/S theo thời gian")
)

fig.add_trace(
    go.Scatter(
        x=bid_cstc["Năm"].astype(str) + "-" + bid_cstc["Tháng"].astype(str),
        y=bid_cstc["P/E"], name="P/E"
    ),
    row=1,
    col=1
)
fig.add_trace(
    go.Scatter(
        x=bid_cstc["Năm"].astype(str) + "-" + bid_cstc["Tháng"].astype(str),
        y=bid_cstc["P/B"], name="P/B"
    ),
    row=2,
    col=1
)
fig.add_trace(
    go.Scatter(
        x=bid_cstc["Năm"].astype(str) + "-" + bid_cstc["Tháng"].astype(str),
        y=bid_cstc["P/S"], name="P/S"
    ),
    row=3,
    col=1
)

fig.update_layout(height=800, width=800, title_text="P/E, P/B, và P/S Ratios")
fig.update_xaxes(tickangle=-45, tickfont=dict(size=10), row=3, col=1)

fig.show()

P/E Ratio Over Time:
- Quý 1/2022 - Quý 2/2023: Tỷ lệ P/E giảm từ khoảng 16 xuống mức thấp nhất là khoảng 10 vào quý 2/2023.
- Quý 2/2023 - 2024: Tỷ lệ P/E có sự hồi phục nhẹ nhưng vào cuối năm 2023 tỷ lệ có sự giảm trở lại.

P/B Ratio Over Time:
- Quý 1/2022 - Quý 2/2022: Tỷ lệ P/B có sự biến động giảm.
- Quý 2/2022 - Quý 1/2023: Tỷ lệ P/B có xu hướng tăng về giá trị đầu năm, sau đó giá trị ổn định quanh mức 2.2.
- Quý 2/2023 - Quý 4/2023: Tỷ lệ P/B có có biến động mạnh, tỷ lệ giảm nhẹ vào giai đoạn đầu, sau đó tăng mạnh vào cuối kì.
- Năm 2024, Tỷ lệ P/B giảm mạnh và đạt đáy vào cuối năm 2024.

P/S Ratio Over Time:
- Tỷ lệ P/S có xu hướng tương đồng với P/B.
- Tỷ lệ P/S có sự biến động đáng kể trong giai đoạn khảo sát, với xu hướng chung là tăng trong giai đoạn đầu và giữa, sau đó có sự sụt giảm nhẹ vào cuối giai đoạn.

In [30]:
fig = go.Figure(data=[go.Bar(
    x=bid_cstc["Năm"].astype(str) + "-" + bid_cstc["Tháng"].astype(str),
    y=bid_cstc["P/Cash Flow"],
    marker_color=["red" if value < 0 else "green" for value in bid_cstc["P/Cash Flow"]]
)])

fig.update_layout(
    title="P/Cash Flow theo thời gian",
    xaxis_title="Năm - Tháng",
    yaxis_title="P/Cash Flow",
    xaxis_tickangle=-45
)

fig.show()

Dựa trên biểu đồ, ta thấy:
- Năm 2022: Tỷ lệ P/Cash Flow có xu hướng giảm nhẹ. Đạt mức thấp nhất là khoảng 1.7.
- Năm 2023: Tỷ lệ P/Cash Flow có sự tăng trưởng vượt bậc vào đầu năm, nhưng đột ngột giảm mạnh xuống mức âm từ quý 2 đến cuối năm.
- Năm 2024: tỷ lệ P/Cash Flow tăng mạnh và đạt đỉnh vào đầu năm, nhưng sau đó có sự điều chỉnh xuống và ổn định.